File: .claude/skills/2c2p-payments/examples/payment-token.js
/**
 * 2C2P PGW v4.3 — Payment Token + Payment Inquiry (Node.js)
 *
 * Demonstrates the JWT (HS256) request/response signing used by all PGW APIs.
 * Wire format is always: { "payload": "<JWT signed with secretKey>" }.
 *
 * Setup:
 *   npm install jsonwebtoken node-fetch    # node-fetch only if Node < 18
 *   export PGW_MERCHANT_ID=your_merchant_id
 *   export PGW_SECRET_KEY=your_secret_key
 *
 * Run:
 *   node payment-token.js
 */

const jwt = require("jsonwebtoken");
// Node 18+ has global fetch. For older Node: const fetch = require("node-fetch");

const MERCHANT_ID = process.env.PGW_MERCHANT_ID;
const SECRET_KEY = process.env.PGW_SECRET_KEY;

// Sandbox base. Swap to https://pgw.2c2p.com for production.
const BASE = "https://sandbox-pgw.2c2p.com/payment/4.3";

if (!MERCHANT_ID || !SECRET_KEY) {
  console.error("Set PGW_MERCHANT_ID and PGW_SECRET_KEY env vars.");
  process.exit(1);
}

/** Sign a payload object as a JWT and POST it; returns the decoded response object. */
async function callPgw(path, payloadObj) {
  const token = jwt.sign(payloadObj, SECRET_KEY, { algorithm: "HS256" });

  const res = await fetch(`${BASE}/${path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ payload: token }),
  });

  const body = await res.json();
  if (!body.payload) {
    throw new Error(`Unexpected response: ${JSON.stringify(body)}`);
  }
  // Verify signature AND decode in one step (throws on tampering).
  return jwt.verify(body.payload, SECRET_KEY, { algorithms: ["HS256"] });
}

async function createPaymentToken() {
  const payload = {
    merchantID: MERCHANT_ID,
    invoiceNo: `INV${Date.now()}`, // unique per order
    description: "Order #1234",
    amount: 100.0,
    currencyCode: "SGD",
    frontendReturnUrl: "https://yourshop.example/return",
    backendReturnUrl: "https://yourshop.example/api/2c2p/callback",
  };

  const resp = await callPgw("paymentToken", payload);
  console.log("respCode:", resp.respCode, "-", resp.respDesc);
  if (resp.respCode === "0000") {
    console.log("Redirect buyer to:", resp.webPaymentUrl);
    console.log("paymentToken:", resp.paymentToken);
    return resp.paymentToken;
  }
  throw new Error(`Payment token failed: ${resp.respCode} ${resp.respDesc}`);
}

async function inquire(invoiceNo) {
  const resp = await callPgw("paymentInquiry", {
    merchantID: MERCHANT_ID,
    invoiceNo,
    locale: "en",
  });
  console.log("Inquiry:", resp.respCode, resp.respDesc, "amount:", resp.amount);
  return resp;
}

createPaymentToken().catch((e) => {
  console.error(e.message);
  process.exit(1);
});

module.exports = { callPgw, createPaymentToken, inquire };

File: .claude/skills/2c2p-payments/examples/payment_token.py
"""
2C2P PGW v4.3 - Payment Token + Payment Inquiry (Python)

Demonstrates the JWT (HS256) request/response signing used by all PGW APIs.
Wire format is always: {"payload": "<JWT signed with secretKey>"}.

Setup:
    pip install pyjwt requests
    export PGW_MERCHANT_ID=your_merchant_id
    export PGW_SECRET_KEY=your_secret_key

Run:
    python payment_token.py
"""

import os
import sys
import time

import jwt  # PyJWT
import requests

MERCHANT_ID = os.environ.get("PGW_MERCHANT_ID")
SECRET_KEY = os.environ.get("PGW_SECRET_KEY")

# Sandbox base. Swap to https://pgw.2c2p.com for production.
BASE = "https://sandbox-pgw.2c2p.com/payment/4.3"

if not MERCHANT_ID or not SECRET_KEY:
    sys.exit("Set PGW_MERCHANT_ID and PGW_SECRET_KEY env vars.")


def call_pgw(path: str, payload: dict) -> dict:
    """Sign payload as JWT, POST it, verify + decode the response JWT."""
    token = jwt.encode(payload, SECRET_KEY, algorithm="HS256")

    resp = requests.post(
        f"{BASE}/{path}",
        json={"payload": token},
        headers={"Content-Type": "application/json"},
        timeout=30,
    )
    resp.raise_for_status()
    body = resp.json()
    if "payload" not in body:
        raise RuntimeError(f"Unexpected response: {body}")

    # Verify signature AND decode (raises on tampering).
    return jwt.decode(body["payload"], SECRET_KEY, algorithms=["HS256"])


def create_payment_token() -> str:
    payload = {
        "merchantID": MERCHANT_ID,
        "invoiceNo": f"INV{int(time.time() * 1000)}",  # unique per order
        "description": "Order #1234",
        "amount": 100.00,
        "currencyCode": "SGD",
        "frontendReturnUrl": "https://yourshop.example/return",
        "backendReturnUrl": "https://yourshop.example/api/2c2p/callback",
    }

    resp = call_pgw("paymentToken", payload)
    print("respCode:", resp.get("respCode"), "-", resp.get("respDesc"))
    if resp.get("respCode") == "0000":
        print("Redirect buyer to:", resp.get("webPaymentUrl"))
        print("paymentToken:", resp.get("paymentToken"))
        return resp.get("paymentToken")
    raise RuntimeError(f"Payment token failed: {resp.get('respCode')} {resp.get('respDesc')}")


def inquire(invoice_no: str) -> dict:
    resp = call_pgw(
        "paymentInquiry",
        {"merchantID": MERCHANT_ID, "invoiceNo": invoice_no, "locale": "en"},
    )
    print("Inquiry:", resp.get("respCode"), resp.get("respDesc"), "amount:", resp.get("amount"))
    return resp


if __name__ == "__main__":
    try:
        create_payment_token()
    except Exception as exc:  # noqa: BLE001
        sys.exit(str(exc))
